File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/vgmrips/Catalog.java
package app.zxtune.fs.vgmrips;

import android.content.Context;

import androidx.annotation.Nullable;

import java.io.IOException;

import app.zxtune.fs.ProgressCallback;
import app.zxtune.fs.http.MultisourceHttpProvider;

public abstract class Catalog {

  public interface Visitor<T> {
    void accept(T obj);
  }

  public interface Grouping {
    void query(Visitor<Group> visitor) throws IOException;
    void queryPacks(String id, Visitor<Pack> visitor, ProgressCallback progress) throws IOException;
  }

  public abstract Grouping companies();
  public abstract Grouping composers();
  public abstract Grouping chips();
  public abstract Grouping systems();

  @Nullable
  public abstract Pack findPack(String id, Visitor<Track> visitor) throws IOException;

  @Nullable
  public abstract Pack findRandomPack(Visitor<Track> visitor) throws IOException;

  public static Catalog create(Context context, MultisourceHttpProvider http) {
    final RemoteCatalog remote = new RemoteCatalog(http);
    final Database db = new Database(context);
    return new CachingCatalog(remote, db);
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/vgmrips/Catalog.kt
package app.zxtune.fs.vgmrips

import android.content.Context
import androidx.core.util.Consumer
import app.zxtune.fs.http.MultisourceHttpProvider
import app.zxtune.utils.ProgressCallback
import java.io.IOException

interface Catalog {

    fun interface Visitor<T> : Consumer<T>

    interface Grouping {
        @Throws(IOException::class)
        fun query(visitor: Visitor<Group>)

        @Throws(IOException::class)
        fun queryPacks(id: Group.Id, visitor: Visitor<Pack>, progress: ProgressCallback)
    }

    fun companies(): Grouping
    fun composers(): Grouping
    fun chips(): Grouping
    fun systems(): Grouping

    @Throws(IOException::class)
    fun findPack(id: Pack.Id): Pack?

    @Throws(IOException::class)
    fun findRandomPack(visitor: Visitor<FilePath>): Pack?

    companion object {
        @JvmStatic
        fun create(context: Context, http: MultisourceHttpProvider): Catalog {
            val remote = RemoteCatalog(http)
            val db = Database(context)
            return CachingCatalog(remote, db)
        }
    }
}
